doc: Adds migration guides to transition out of Serverless and Shared-tier clusters (#2746)
mongodb/terraform-provider-mongodbatlas: docs/guides/serverless-shared-migration-guide.md

---
page_title: "Migration Guide: Transition out of Serverless Instances and Shared-tier clusters"
---

> 

Reviewer: Noting that migration guide for Flex-to-Dedicated will be included as a part of [CLOUDP-278908](https://jira.mongodb.org/browse/CLOUDP-278908), as noted [here](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2744/files#r1820614013).

# Migration Guide: Transition out of Serverless Instances and Shared-tier clusters

> The goal of this guide is to help users transition out of Serverless Instances (`mongodbatlas_serverless_instance`) and Shared-tier clusters (M2 and M5) in favor of Flex or Dedicated Clusters. 

Reviewer 2: ```suggestion
The goal of this guide is to help users transition from Serverless Instances (`mongodbatlas_serverless_instance`) and Shared-tier clusters (M2 and M5) to Flex or Dedicated Clusters. 
```

The 'from'-'to' relationship calls less attention to itself than 'out of' and 'in favor of'.

> After March 2025, all Serverless instances and Shared-tier clusters will be automatically converted to Flex/Dedicated clusters, results in configuration drift in Terraform. If a Serverless instance/Shared-tier cluster does not fit the constraints of a Flex cluster, it will be converted into a Dedicated cluster which will result in downtime and workload disruption. It will otherwise be converted into a Flex cluster.

Reviewer 2: ```suggestion
After March 2025, all Serverless instances and Shared-tier clusters will automatically convert to Flex/Dedicated clusters, resulting in configuration drift in Terraform. If a Serverless instance or Shared-tier cluster does not fit the constraints of a Flex cluster, it will convert into a Dedicated cluster, resulting in downtime and workload disruption. Otherwise, it will convert to a Flex cluster.
```
The difference between passive and active is not technically substantive in this case, so prefer the active voice.

Also, a question: is 'configuration drift' a well-established term? If not, we should consider rephrasing this.

Alternatively, we could structure this like so:

```suggestion
After March 2025, all Serverless instances and Shared-tier clusters will automatically convert to either:

- Flex clusters, if they meet the relevant operating constraints
- Dedicated clusters, if they do not meet Flex cluster constraints; in this scenario, expect downtime and 
  workload disruption

This conversion results in configuration drift in Terraform
```

The relationships between the key bits of information are a bit more explicit this way.

> Migration from Serverless instances and Shared-tier clusters to Flex/Dedicated clusters can be done manually before March 2025. 

Reviewer 2: ```suggestion
You can migrate from Serverless instances and Shared-tier clusters to Flex or Dedicated clusters manually before March 2025. 
```

Favor active voice over passive voice.

> **Note:** We recommend to wait until March 2025 for Serverless instances and Shared-tier clusters to be autoconverted.

Reviewer 2: ```suggestion
**Note:** We recommend waiting until March 2025 for Serverless instances and Shared-tier clusters to autoconvert.
```

Favor active voice over passive voice.

## From M2/M5 clusters to Flex 
### Pre-Autoconversion Migration Procedure

1. Create a new Flex Cluster directly from your `.tf` file, e.g.:

    ```terraform
    resource "mongodbatlas_flex_cluster" "flex_cluster" {
        project_id = var.project_id
        name       = "flexClusterName"
        provider_settings = {
            backing_provider_name = "AWS"
            region_name           = "US_EAST_1"
        }
        termination_protection_enabled = true
    }
    ```
2. Run `terraform apply` to create the new resource.
3. Migrate data from your Shared-tier cluster to the Flex cluster using `mongodump` and `mongostore`.

    Please see the following guide on how to retrieve data from one cluster and store it in another cluster: [Convert a Serverless Instance to a Dedicated Cluster](https://www.mongodb.com/docs/atlas/tutorial/convert-serverless-to-dedicated/)

    Verify that your data is present within the Flex cluster before proceeding.

> 4. Delete the Shared-tier cluster by running a destory command targetting the Shared-tier cluster:

Reviewer 2: ```suggestion
4. Delete the Shared-tier cluster by running a destroy command against it:
```
There is no other possible referent, so 'it' can be safely used here.

    `terraform destroy -target=mongodbatlas_advanced_cluster.<shared-tier-cluster-name>`

>  5. You may now safely remove the resource block for the Shared-tier cluster from your `.tf` file.

Reviewer 2: ```suggestion
 5. Remove the resource block for the Shared-tier cluster from your `.tf` file.
```
Unless we have a realistic expectation that a user may try to preempt this instruction at an inappropriate time, I think we can cut the safety statement. If that is a realistic expectation, a warning before the procedure is probably more effective than this statement at the end.

### Post-Autoconversion Migration Procedure

> Given your Shared-tier cluster fits the constraints of a Flex cluster, it will automatically convert to a Flex cluster in Atlas, retaining all data. The following resolves the configuration drift in Terraform:

Reviewer 3: > `it will automatically` 

Can we be more specific with a month here?

> Given your Shared-tier cluster fits the constraints of a Flex cluster, it will automatically convert to a Flex cluster in Atlas, retaining all data. The following resolves the configuration drift in Terraform:

Reviewer 3: ```suggestion
Given your Shared-tier cluster fits the constraints of a Flex cluster, it will automatically convert to a Flex cluster in Atlas, retaining all data. 
The following resolves the configuration drift in Terraform and does not affect the underlying cluster infrastructure:
```

> Given your Shared-tier cluster fits the constraints of a Flex cluster, it alongisde all its data will have been automatically converted into a Flex cluster in Atlas. The following will resolve the configuration drift in Terraform.

Reviewer 2: ```suggestion
Given your Shared-tier cluster fits the constraints of a Flex cluster, it will automatically convert to a Flex cluster in Atlas, retaining all data. The following resolves the configuration drift in Terraform:
```
Future tense is okay in the first instance, but let's still prefer the active voice. In the second instance, there is no reason it needs to be in future tense.

1. Find the import IDs of the Flex clusters: `{PROJECT_ID}-{CLUSTER_NAME}`, such as `664619d870c247237f4b86a6-flexClusterName`
2. Add an import block per cluster to one of your `.tf` files:
    ```terraform
    import {
    to = mongodbatlas_flex_cluster.flex_cluster
    id = "664619d870c247237f4b86a6-flexClusterName" # from step 1
    }
    ```
3. Run `terraform plan -generate-config-out=flex_cluster.tf`. This should generate a `flex_cluster.tf` file with your Flex cluster in it.
4. Run `terraform apply`. You should see the resource(s) imported: `Apply complete! Resources: 1 imported, 0 added, 0 changed, 0 destroyed.`

> 5. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration will have static values. Look in your previous configuration for:

Reviewer 2: ```suggestion
5. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration have static values. Look in your previous configuration for:
```

Favor present tense.

   - variables, for example: `var.project_id`
   - Terraform keywords, for example: `for_each`, `count`, and `depends_on`

> 6. Replace your existing clusters with the ones from `flex_cluster.tf` and run `terraform state rm mongodbatlas_advanced_cluster.advClusterName`. Without this step, Terraform will create a plan to delete your existing cluster.

Reviewer 2: ```suggestion
6. Replace your existing clusters with the ones from `flex_cluster.tf` and run `terraform state rm mongodbatlas_advanced_cluster.advClusterName`. Without this step, Terraform creates a plan to delete your existing cluster.
```
Favor present tense.

7.  Remove the import block created in step 2.

> 8.  Re-run `terraform apply` to ensure you have no planned changes: `No changes. Your infrastructure matches the configuration.` 

Reviewer: Noting these steps follow closely steps outlined in [cluster-to-advanced-cluster-migration-guide.md](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/docs/guides/cluster-to-advanced-cluster-migration-guide.md?plain=1#L94-L180).

## From Serverless to Flex

### Pre-Autoconversion Migration Procedure

1. Create a new Flex Cluster directly from your `.tf` file, e.g.:

    ```terraform
    resource "mongodbatlas_flex_cluster" "flex_cluster" {
        project_id = var.project_id
        name       = "flexClusterName"
        provider_settings = {
            backing_provider_name = "AWS"
            region_name           = "US_EAST_1"
        }
        termination_protection_enabled = true
    }
    ```
2.  Run `terraform apply` to create the new resource.
3. Migrate data from your Serverless Instance to the Flex cluster using `mongodump` and `mongostore`.

    Please see the following guide on how to retrieve data from one cluster and store it in another cluster: [Convert a Serverless Instance to a Dedicated Cluster](https://www.mongodb.com/docs/atlas/tutorial/convert-serverless-to-dedicated/)

    Verify that your data is present within the Flex cluster before proceeding.

> 4. Delete the Serverless Instance by running a destory command targetting the Serverless Instance:

Reviewer 2: ```suggestion
4. Delete the Serverless Instance by running a destroy command against it:
```
Same comment for the analogous line in the previous procedure.

    `terraform destroy -target=mongodbatlas_serverless_instance.<serverless-instance-cluster-name>`

 5. You may now safely remove the resource block for the Serverless Instance from your `.tf` file.

### Post-Autoconversion Migration Procedure

> Given your Serverless Instance fits the constraints of a Flex cluster, it alongisde all its data will have been automatically converted into a Flex cluster in Atlas. The following will resolve the configuration drift in Terraform.

Reviewer 2: ```suggestion
Given your Serverless Instance fits the constraints of a Flex cluster, it will automatically convert into a Flex cluster in Atlas, retaining all data. The following resolves the configuration drift in Terraform.
```
Same comments as for the analogous line.

1. Find the import IDs of the Flex clusters: `{PROJECT_ID}-{CLUSTER_NAME}`, such as `664619d870c247237f4b86a6-flexClusterName`
2. Add an import block per cluster to one of your `.tf` files:
    ```terraform
    import {
    to = mongodbatlas_flex_cluster.flex_cluster
    id = "664619d870c247237f4b86a6-flexClusterName" # from step 1
    }
    ```
3. Run `terraform plan -generate-config-out=flex_cluster.tf`. This should generate a `flex_cluster.tf` file with your Flex cluster in it.
4. Run `terraform apply`. You should see the resource(s) imported: `Apply complete! Resources: 1 imported, 0 added, 0 changed, 0 destroyed.`

> 5. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration will have static values. Look in your previous configuration for:

Reviewer 2: ```suggestion
5. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration have static values. Look in your previous configuration for:
```
Favor present tense.

   - variables, for example: `var.project_id`
   - Terraform keywords, for example: `for_each`, `count`, and `depends_on`

> 6. Replace your existing clusters with the ones from `flex_cluster.tf` and run `terraform state rm mongodbatlas_serverless_instance.serverlessInstanceName`. Without this step, Terraform will create a plan to delete your existing cluster.

Reviewer 2: ```suggestion
6. Replace your existing clusters with the ones from `flex_cluster.tf` and run `terraform state rm mongodbatlas_serverless_instance.serverlessInstanceName`. Without this step, Terraform creates a plan to delete your existing cluster.
```
Favor present tense.

7.  Remove the import block created in step 2.
8.  Re-run `terraform apply` to ensure you have no planned changes: `No changes. Your infrastructure matches the configuration.` 

## From Serverless to Dedicated 

> Migration from Serverless to Dedicated cannot be done using the Terraform provider.

Reviewer 2: ```suggestion
You cannot migrate from Serverless to Dedicated using the Terraform provider.
```

> **Note:** We recommend waiting until January 2025 for a UI based migration tool.

Reviewer 2: ```suggestion
**Note:** We recommend waiting until January 2025 for a UI-based migration tool.
```

### Pre-Autoconversion Migration Procedure

To migrate from Serverless to Dedicated prior to January 2025, please see the following guide: [Convert a Serverless Instance to a Dedicated Cluster](https://www.mongodb.com/docs/atlas/tutorial/convert-serverless-to-dedicated/)

### Post-Autoconversion Migration Procedure

> Given your Serverless Instance did not fit the constraints of a Flex cluster, it alongisde all its data will have been automatically converted into a Dedicated cluster in Atlas. The following will resolve the configuration drift in Terraform.

Reviewer 2: ```suggestion
Given your Serverless Instance doesn't fit the constraints of a Flex cluster, it will automatically convert into a Dedicated cluster in Atlas, retaining all data. The following resolves the configuration drift in Terraform:
```
Favor present tense, favor active voice.

1. Find the import IDs of the Dedicated clusters: `{PROJECT_ID}-{CLUSTER_NAME}`, such as `664619d870c247237f4b86a6-advancedClusterName`
2. Add an import block per cluster to one of your `.tf` files:
    ```terraform
    import {
    to = mongodbatlas_advanced_cluster.advanced_cluster
    id = "664619d870c247237f4b86a6-advancedClusterName" # from step 1
    }
    ```
3. Run `terraform plan -generate-config-out=adv_cluster.tf`. This should generate a `adv_cluster.tf` file with your Dedicated cluster in it.
4. Run `terraform apply`. You should see the resource(s) imported: `Apply complete! Resources: 1 imported, 0 added, 0 changed, 0 destroyed.`

> 5. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration will have static values. Look in your previous configuration for:

Reviewer 2: ```suggestion
5. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration have static values. Look in your previous configuration for:
```
Favor present tense.

   - variables, for example: `var.project_id`
   - Terraform keywords, for example: `for_each`, `count`, and `depends_on`

> 6. Replace your existing clusters with the ones from `adv_cluster.tf` and run `terraform state rm mongodbatlas_serverless_instance.serverlessInstanceName`. Without this step, Terraform will create a plan to delete your existing cluster.

Reviewer 2: ```suggestion
6. Replace your existing clusters with the ones from `adv_cluster.tf` and run `terraform state rm mongodbatlas_serverless_instance.serverlessInstanceName`. Without this step, Terraform creates a plan to delete your existing cluster.
```
Favor present tense.

## From Serverless to Free 

### Pre-Autoconversion Migration Procedure

1. Create a new Free Cluster directly from your `.tf` file, e.g.:

    ```terraform
    resource "mongodbatlas_advanced_cluster" "free-cluster" {
        project_id   = var.atlas_project_id
        name         = "freeClusterName"
        cluster_type = "REPLICASET"

        replication_specs {
            region_configs {
            electable_specs {
                instance_size = "M0"
            }
            provider_name         = "TENANT"
            backing_provider_name = "AWS"
            region_name           = "US_EAST_1"
            priority              = 7
            }
        }
    }
    ```
2.  Run `terraform apply` to create the new resource.
3. Migrate data from your Serverless Instance to the Free cluster using `mongodump` and `mongostore`.

    Please see the following guide on how to retrieve data from one cluster and store it in another cluster: [Convert a Serverless Instance to a Dedicated Cluster](https://www.mongodb.com/docs/atlas/tutorial/convert-serverless-to-dedicated/)

    Verify that your data is present within the Free cluster before proceeding.

> 4. Delete the Serverless Instance by running a destory command targetting the Serverless Instance:

Reviewer 2: ```suggestion
4. Delete the Serverless Instance by running a destroy command against the Serverless Instance:
```
Typo fix, wording suggestion

    `terraform destroy -target=mongodbatlas_serverless_instance.<serverless-instance-cluster-name>`

>  5. You may now safely remove the resource block for the Serverless Instance from your `.tf` file.

Reviewer 2: ```suggestion
 5. Remove the resource block for the Serverless Instance from your `.tf` file.
```
Same comment as on analogous line.

### Post-Autoconversion Migration Procedure

> Given your Serverless Instance has $0 MRR, it alongisde all its data will have been automatically converted into a Free cluster in Atlas. The following will resolve the configuration drift in Terraform.

Reviewer 2: ```suggestion
Given your Serverless Instance has $0 MRR, it automatically converts into a Free cluster in Atlas, retaining all data. The following resolves the configuration drift in Terraform.
```
Same comments as on analogous lines.

1. Find the import IDs of the Free clusters: `{PROJECT_ID}-{CLUSTER_NAME}`, such as `664619d870c247237f4b86a6-freeClusterName`
2. Add an import block per cluster to one of your `.tf` files:
    ```terraform
    import {
    to = mongodbatlas_advanced_cluster.free_cluster
    id = "664619d870c247237f4b86a6-freeClusterName" # from step 1
    }
    ```
3. Run `terraform plan -generate-config-out=free_cluster.tf`. This should generate a `free_cluster.tf` file with your Free cluster in it.
4. Run `terraform apply`. You should see the resource(s) imported: `Apply complete! Resources: 1 imported, 0 added, 0 changed, 0 destroyed.`

> 5. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration will have static values. Look in your previous configuration for:

Reviewer 2: ```suggestion
5. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration have static values. Look in your previous configuration for:
```
Favor present tense.

   - variables, for example: `var.project_id`
   - Terraform keywords, for example: `for_each`, `count`, and `depends_on`

> 6. Replace your existing clusters with the ones from `free_cluster.tf` and run `terraform state rm mongodbatlas_serverless_instance.serverlessInstanceName`. Without this step, Terraform will create a plan to delete your existing cluster.

Reviewer 2: ```suggestion
6. Replace your existing clusters with the ones from `free_cluster.tf` and run `terraform state rm mongodbatlas_serverless_instance.serverlessInstanceName`. Without this step, Terraform creates a plan to delete your existing cluster.
```

---
page_title: "Migration Guide: Transition out of Serverless Instances and Shared-tier clusters"
---

# Migration Guide: Transition out of Serverless Instances and Shared-tier clusters

The goal of this guide is to help users transition from Serverless Instances and Shared-tier clusters (M2/M5) to Free, Flex or Dedicated Clusters. 

> After March 2025, all Shared-tier clusters (`mongodbatlas_cluster` and `mongodbatlas_advanced_cluster`) will automatically convert to Flex clusters and all Serverless instances (`mongodb_serverless_instance`) will be converted into Free/Flex/Dedicated clusters, resulting in configuration drift in Terraform. 

Reviewer 3: ```suggestion
After March 2025, all Shared-tier clusters (in both `mongodbatlas_cluster` and `mongodbatlas_advanced_cluster`) will automatically convert to Flex clusters. Similarly, all Serverless instances (`mongodb_serverless_instance`) will be converted into Free/Flex/Dedicated clusters, depending on your existing configuration.
These conversions will result in configuration drift in Terraform. 
```

Reviewer 3: for the "depending on your existing configuration" can we please find an atlas link that will explain how the conversion will happen?

Reviewer: Added placeholder link for the moment while trying to obtain link from upstream 👍

You can migrate from Serverless instances and Shared-tier clusters manually before March 2025. 

> **NOTE:** We recommend waiting until March 2025 for Serverless instances and Shared-tier clusters to autoconvert.

Reviewer 3: can we repeat this into every "pre-autoconversion" section?

Reviewer: Done

If a Serverless instance has $0 MRR, it automatically converts into a Free cluster. Else, if it does not fit the constraints of a Flex cluster, it will convert into a Dedicated cluster, resulting in downtime and workload disruption. Otherwise, it will convert to a Flex cluster.

> We are working on enhancing the User Experience such that, in future, Terraform Atlas Providers users can make even fewer required changes to their scripts from what is shown below. More updates to come over the next few months. For more details reach out to: zuhair.ahmed@mongodb.com

Reviewer 3: ```suggestion
For Shared-tier clusters, we are working on enhancing the User Experience such that Terraform Atlas Providers users can make even fewer required changes to their scripts from what is shown below. More updates to come over the next few months. For more details reach out to: zuhair.ahmed@mongodb.com
```

5. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration have static values. Look in your previous configuration for:
   - variables, for example: `var.project_id`
   - Terraform keywords, for example: `for_each`, `count`, and `depends_on`
6. Update the references from your previous cluster resource: `mongodbatlas_advanced_cluster.this.X` or `mongodbatlas_cluster.this.X` to the new `mongodbatlas_flex_cluster.this.X`.
7. Update any shared-tier data source blocks to refer to `mongodbatlas_flex_cluster`.
8. Replace your existing clusters with the ones from `flex_cluster.tf` and run 

    `terraform state rm mongodbatlas_advanced_cluster.this`

    or `terraform state rm mongodbatlas_cluster.this`. 

    Without this step, Terraform creates a plan to delete your existing cluster.

9.  Remove the import block created in step 2.
10.  Re-run `terraform apply` to ensure you have no planned changes: `No changes. Your infrastructure matches the configuration.` 

## From Serverless to Free 

### Pre-Autoconversion Migration Procedure

> 

Reviewer 3: can we also add a note to do this only if you've verified your serverless cluster can fit into a free cluster?

Reviewer: Added requirements for all three Serverless migration guides 👍

---
page_title: "Migration Guide: Transition out of Serverless Instances and Shared-tier clusters"
---

# Migration Guide: Transition out of Serverless Instances and Shared-tier clusters

The goal of this guide is to help users transition from Serverless Instances and Shared-tier clusters (M2/M5) to Free, Flex or Dedicated Clusters. 

Starting in March 2025, all Shared-tier clusters (in both `mongodbatlas_cluster` and `mongodbatlas_advanced_cluster`) will automatically convert to Flex clusters. Similarly, all Serverless instances (`mongodb_serverless_instance`) will be converted into Free/Flex/Dedicated clusters, [depending on your existing configuration](https://www.mongodb.com/docs/atlas/flex-migration/). 

> These conversions will result in configuration drift in Terraform. 

Reviewer 3: not always right? shared tier migrates to flex but still works as expected

Reviewer 3: I believe cases where configuration drift occurs (in March 2025 or whenever auto-migration is done) are when serverless is converted to either free or dedicated.
When serverless or shared tier get converted to flex, API endpoints keep working until January 2026

Author: changed a bit the phrasing in https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2746/commits/b6944698fdb057950b42b5163edcda5449344094, also changed the date of shared tier autoconversion to January, which is what we have in the migrtion doc from the upstream team

> These conversions will result in configuration drift in Terraform. 

Reviewer 3: and also we should make clear that by January 2026 they wil stop working

Author: As per my understanding, Serverless and Shared-tier will be deprecated in January 2025, and EOL by Jan 2026, and we should share that serverless and shared tier will stop working in January 2025 along the deprecations. If we share this here, I think it might be confusing because the resources will not be deprecated yet

You can migrate from Serverless instances and Shared-tier clusters manually before March 2025. 

**--> NOTE:** We recommend waiting until March 2025 or later for Serverless instances and Shared-tier clusters to autoconvert.

If a Serverless instance has $0 MRR, it automatically converts into a Free cluster. Else, if it does not fit the constraints of a Flex cluster, it will convert into a Dedicated cluster, resulting in downtime and workload disruption. Otherwise, it will convert to a Flex cluster.

For Shared-tier clusters, we are working on enhancing the User Experience such that Terraform Atlas Providers users can make even fewer required changes to their scripts from what is shown below. More updates to come over the next few months. For more details reach out to: zuhair.ahmed@mongodb.com

### Jump to:
- [Shared-tier to Flex](#from-shared-tier-clusters-to-flex)
- [Serverless to Free](#from-serverless-to-free)
- [Serverless to Flex](#from-serverless-to-flex)
- [Serverless to Dedicated](#from-serverless-to-dedicated)

## From Shared-tier clusters to Flex 
### Pre-Autoconversion Migration Procedure

> **NOTE:** We recommend waiting until March 2025 or later for Shared-tier clusters to autoconvert.

Reviewer 3: doesn't this cause downtime? should that be clear?

Author: I am not aware of this auto conversion to cause downtime, at least I haven't seen it in any docs. The one auto conversion that will cause downtime that I know of is Serverless to Dedicated

Reviewer 3: I am referrring to when they do it manually, if they are migrating manually shouldn't we warn them it causes downtime? It requires creating a new cluster and migrate data..

Author: RIght, I made it explicit in all the pre-autoconvestion sections in this commit https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2746/commits/c5f4719ed46345e27e0ca0a2880c6bd66cdb21b6

    ```terraform
    resource "mongodbatlas_flex_cluster" "this" {
        project_id = var.project_id
        name       = "flexClusterName"
        provider_settings = {
            backing_provider_name = "AWS"
            region_name           = "US_EAST_1"
        }
        termination_protection_enabled = true
    }
    ```
2. Run `terraform apply` to create the new resource.
3. Migrate data from your Shared-tier cluster to the Flex cluster using `mongodump` and `mongostore`.

    Please see the following guide on how to retrieve data from one cluster and store it in another cluster: [Convert a Serverless Instance to a Dedicated Cluster](https://www.mongodb.com/docs/atlas/tutorial/convert-serverless-to-dedicated/)

    Verify that your data is present within the Flex cluster before proceeding.
4. Delete the Shared-tier cluster by running a destroy command against it.
    
    For *mongodbatlas_advanced_cluster*:

    `terraform destroy -target=mongodbatlas_advanced_cluster.this`

    For *mongodbatlas_cluster*:

    `terraform destroy -target=mongodbatlas_cluster.this`

 5. Remove the resource block for the Shared-tier cluster from your `.tf` file.

> ### Post-Autoconversion Migration Procedure

Reviewer 3: should this go first?

Author: Done

6. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration have static values. Look in your previous configuration for:
   - variables, for example: `var.project_id`
   - Terraform keywords, for example: `for_each`, `count`, and `depends_on`
7. Update the references from your previous cluster resource: `mongodbatlas_advanced_cluster.this.X` or `mongodbatlas_cluster.this.X` to the new `mongodbatlas_flex_cluster.this.X`.
8. Update any shared-tier data source blocks to refer to `mongodbatlas_flex_cluster`.
9. Replace your existing clusters with the ones from `flex_cluster.tf` and run 

    `terraform state rm mongodbatlas_advanced_cluster.this`

    or `terraform state rm mongodbatlas_cluster.this`. 

    Without this step, Terraform creates a plan to delete your existing cluster.

10.  Remove the import block created in step 2.
11.  Re-run `terraform plan` to ensure you have no planned changes: `No changes. Your infrastructure matches the configuration.` 

## From Serverless to Free 

> **Please ensure your Serverless instance meets the following requirements to migrate to Free:**

Reviewer 3: official doc?

Reviewer: To my understanding (from [this conversation with upstream](https://mongodb.slack.com/archives/C07BU4RHNJK/p1731081287931629?thread_ts=1731068246.540079&cid=C07BU4RHNJK)), there won't be any official doc with migration requirement info.

Author: Also, there's no explicit mention of constraints (memory, ops sec, etc) in the official docs https://www.mongodb.com/docs/atlas/flex-migration/ it is mentioned “To see which tiers Atlas will migrate your instances to, consult the [All Clusters](https://cloud.mongodb.com/v2#/clusters) page in the Atlas UI.”, and for that the customer will need to access the UI, which some customers can’t do/don’t want to do if they are using TF

5. Remove the "default" fields. Many fields of this resource are optional. Look for fields with a `null` or `0` value.
6. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration have static values. Look in your previous configuration for:
   - variables, for example: `var.project_id`
   - Terraform keywords, for example: `for_each`, `count`, and `depends_on`
7. Update the references from your previous cluster resource: `mongodbatlas_serverless_instance.this.X` to the new `mongodbatlas_advanced_cluster.this.X`.
8. Update any shared-tier data source blocks to refer to `mongodbatlas_advanced_cluster`.
9. Replace your existing clusters with the ones from `free_cluster.tf` and run `terraform state rm mongodbatlas_serverless_instance.this`. Without this step, Terraform creates a plan to delete your existing cluster.
10.  Remove the import block created in step 2.
11.  Re-run `terraform plan` to ensure you have no planned changes: `No changes. Your infrastructure matches the configuration.` 

## From Serverless to Flex

> **Please ensure your Serverless instance meets the following requirements to migrate to Flex:**

Reviewer 3: is it possible to just link official doc here?

5. Remove the "default" fields. Many fields of this resource are optional. Look for fields with a `null` or `0` value.
6. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration have static values. Look in your previous configuration for:
   - variables, for example: `var.project_id`
   - Terraform keywords, for example: `for_each`, `count`, and `depends_on`
7. Update the references from your previous cluster resource: `mongodbatlas_serverless_instance.this.X` to the new `mongodbatlas_flex_cluster.this.X`.
8. Update any shared-tier data source blocks to refer to `mongodbatlas_flex_cluster`.
9. Replace your existing clusters with the ones from `flex_cluster.tf` and run `terraform state rm mongodbatlas_serverless_instance.this`. Without this step, Terraform creates a plan to delete your existing cluster.
10.  Remove the import block created in step 2.
11.  Re-run `terraform plan` to ensure you have no planned changes: `No changes. Your infrastructure matches the configuration.` 

## From Serverless to Dedicated 

> **Please note your Serverless instance will need to migrate to Decidated if it meets the following requirements:**

Reviewer 3: official doc

- \>= 5GB of data
- needs privatelink or continuous backup
- \> 500 ops/sec consistently.

You cannot migrate from Serverless to Dedicated using the Terraform provider.

> ### Pre-Autoconversion Migration Procedure

Reviewer 3: same as before: post- before pre- since post- is recommended

Author: Done

**NOTE:** We recommend waiting until March 2025 or later for Serverless instances to autoconvert.

In early 2025, a UI-based migration tool will be released to assist you in migrating your workload prior to the autoconversion.

To migrate from Serverless to Dedicated prior to early 2025, please see the following guide: [Convert a Serverless Instance to a Dedicated Cluster](https://www.mongodb.com/docs/atlas/tutorial/convert-serverless-to-dedicated/)

### Post-Autoconversion Migration Procedure

> Given your Serverless Instance doesn't fit the constraints of a Flex cluster, it will automatically convert in March 2025 into a Dedicated cluster in Atlas, retaining all data.

Reviewer 3: specify this guide is only valid after auto-conversion is done, now the text mentions it but I think it's not clear

Author: Let me know if it's clearer now

---
page_title: "Migration Guide: Transition out of Serverless Instances and Shared-tier clusters"
---

# Migration Guide: Transition out of Serverless Instances and Shared-tier clusters

The goal of this guide is to help users transition from Serverless Instances and Shared-tier clusters (M2/M5) to Free, Flex or Dedicated Clusters. 

Starting in January 2025, all Shared-tier clusters (in both `mongodbatlas_cluster` and `mongodbatlas_advanced_cluster`) will automatically convert to Flex clusters. Similarly, in March 2025 all Serverless instances (`mongodb_serverless_instance`) will be converted into Free/Flex/Dedicated clusters, [depending on your existing configuration](https://www.mongodb.com/docs/atlas/flex-migration/).
If a Serverless instance has $0 MRR, it automatically converts into a Free cluster. Else, if it does not fit the constraints of a Flex cluster, it will convert into a Dedicated cluster, resulting in downtime and workload disruption. Otherwise, it will convert to a Flex cluster.
Some of these conversions will result in configuration drift in Terraform. 

> You can migrate from Serverless instances and Shared-tier clusters manually before March 2025. 

Reviewer: Since dates of autoconversion are different for serverless and m2/m5, it may be better to leave out a date
```suggestion
You can migrate from Serverless instances and Shared-tier clusters manually before autoconversion. 
```

6. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration have static values. Look in your previous configuration for:
   - variables, for example: `var.project_id`
   - Terraform keywords, for example: `for_each`, `count`, and `depends_on`
7. Update the references from your previous cluster resource: `mongodbatlas_advanced_cluster.this.X` or `mongodbatlas_cluster.this.X` to the new `mongodbatlas_flex_cluster.this.X`.
8. Update any shared-tier data source blocks to refer to `mongodbatlas_flex_cluster`.
9. Replace your existing clusters with the ones from `flex_cluster.tf` and run 

    `terraform state rm mongodbatlas_advanced_cluster.this`

    or `terraform state rm mongodbatlas_cluster.this`. 

    Without this step, Terraform creates a plan to delete your existing cluster.

10.  Remove the import block created in step 2.
11.  Re-run `terraform plan` to ensure you have no planned changes: `No changes. Your infrastructure matches the configuration.` 

### Pre-Autoconversion Migration Procedure

> **NOTE:** We recommend waiting until March 2025 or later for Shared-tier clusters to autoconvert. Manually doing the migration can cause downtime and workload disruption.

Reviewer: ```suggestion
**NOTE:** We recommend waiting until January 2025 or later for Shared-tier clusters to autoconvert. Manually doing the migration can cause downtime and workload disruption.
```

    }
    ```
2. Run `terraform apply` to create the new resource.
3. Migrate data from your Shared-tier cluster to the Flex cluster using `mongodump` and `mongostore`.

    Please see the following guide on how to retrieve data from one cluster and store it in another cluster: [Convert a Serverless Instance to a Dedicated Cluster](https://www.mongodb.com/docs/atlas/tutorial/convert-serverless-to-dedicated/)

    Verify that your data is present within the Flex cluster before proceeding.
4. Delete the Shared-tier cluster by running a destroy command against it.
    
    For *mongodbatlas_advanced_cluster*:

    `terraform destroy -target=mongodbatlas_advanced_cluster.this`

    For *mongodbatlas_cluster*:

    `terraform destroy -target=mongodbatlas_cluster.this`

 5. Remove the resource block for the Shared-tier cluster from your `.tf` file.

## From Serverless to Free 

**Please ensure your Serverless instance meets the following requirements to migrate to Free:**
- $0 MRR

### Post-Autoconversion Migration Procedure

Given your Serverless Instance has $0 MRR, it will automatically convert in March 2025 into a Free cluster in Atlas, retaining all data.

> The following steps resolve the configuration drift in Terraform and does not affect the underlying cluster infrastructure:

Reviewer 2: ```suggestion
The following steps resolve the configuration drift in Terraform without affecting the underlying cluster infrastructure:
```
Edit for flow.

2. Add an import block per cluster to one of your `.tf` files:
    ```terraform
    import {
    to = mongodbatlas_advanced_cluster.this
    id = "664619d870c247237f4b86a6-freeClusterName" # from step 1
    }
    ```
3. Run `terraform plan -generate-config-out=free_cluster.tf`. This should generate a `free_cluster.tf` file with your Free cluster in it.
4. Run `terraform apply`. You should see the resource(s) imported: `Apply complete! Resources: 1 imported, 0 added, 0 changed, 0 destroyed.`
5. Remove the "default" fields. Many fields of this resource are optional. Look for fields with a `null` or `0` value.
6. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration have static values. Look in your previous configuration for:
   - variables, for example: `var.project_id`
   - Terraform keywords, for example: `for_each`, `count`, and `depends_on`
7. Update the references from your previous cluster resource: `mongodbatlas_serverless_instance.this.X` to the new `mongodbatlas_advanced_cluster.this.X`.
8. Update any shared-tier data source blocks to refer to `mongodbatlas_advanced_cluster`.

> 9. Replace your existing clusters with the ones from `free_cluster.tf` and run `terraform state rm mongodbatlas_serverless_instance.this`. Without this step, Terraform creates a plan to delete your existing cluster.

Reviewer 4: Have we considered the `remove`+`import` solution instead?

5. Remove the "default" fields. Many fields of this resource are optional. Look for fields with a `null` or `0` value.
6. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated configuration have static values. Look in your previous configuration for:
   - variables, for example: `var.project_id`
   - Terraform keywords, for example: `for_each`, `count`, and `depends_on`
7. Update the references from your previous cluster resource: `mongodbatlas_serverless_instance.this.X` to the new `mongodbatlas_flex_cluster.this.X`.
8. Update any shared-tier data source blocks to refer to `mongodbatlas_flex_cluster`.
9. Replace your existing clusters with the ones from `flex_cluster.tf` and run `terraform state rm mongodbatlas_serverless_instance.this`. Without this step, Terraform creates a plan to delete your existing cluster.
10.  Remove the import block created in step 2.
11.  Re-run `terraform plan` to ensure you have no planned changes: `No changes. Your infrastructure matches the configuration.` 

### Pre-Autoconversion Migration Procedure

> **NOTE:** We recommend waiting until March 2025 or later for Serverless instances to autoconvert. Manually doing the migration can cause downtime and workload disruption.

Reviewer 2: ```suggestion
**NOTE:** We recommend waiting until March 2025 or later for Serverless instances to autoconvert. Manual migration can cause downtime and workload disruption.
```
Edit for conciseness.

        }
        termination_protection_enabled = true
    }
    ```
2.  Run `terraform apply` to create the new resource.
3. Migrate data from your Serverless Instance to the Flex cluster using `mongodump` and `mongostore`.

    Please see the following guide on how to retrieve data from one cluster and store it in another cluster: [Convert a Serverless Instance to a Dedicated Cluster](https://www.mongodb.com/docs/atlas/tutorial/convert-serverless-to-dedicated/)

    Verify that your data is present within the Flex cluster before proceeding.
4. Delete the Serverless Instance by running a destroy command against it:

    `terraform destroy -target=mongodbatlas_serverless_instance.this`

 5. You may now safely remove the resource block for the Serverless Instance from your `.tf` file.

## From Serverless to Dedicated 

**Please note your Serverless instance will need to migrate to Decidated if it meets the following requirements:**
- \>= 5GB of data
- needs privatelink or continuous backup
- \> 500 ops/sec consistently.

You cannot migrate from Serverless to Dedicated using the Terraform provider.

### Pre-Autoconversion Migration Procedure

> **NOTE:** In early 2025, a UI-based tool will be released to assist you in migrating your workloads from Serverless instances to Dedicated clusters. This tool will ensure correct migration and you will not need to change connection strings. There will be some downtime while using this tool.

Reviewer 2: ```suggestion
**NOTE:** In early 2025, we will release a UI-based tool for migrating your workloads from Serverless instances to Dedicated clusters. This tool will ensure correct migration with little downtime. You won't need to change connection strings.
```
This is not only a little tighter, it approaches the downtime issue from a more positive direction.

> To migrate from Serverless to Dedicated prior to early 2025, please see the following guide: [Convert a Serverless Instance to a Dedicated Cluster](https://www.mongodb.com/docs/atlas/tutorial/convert-serverless-to-dedicated/). **NOTE:** Manually doing the migration can cause downtime and workload disruption.

Reviewer 2: ```suggestion
To migrate from Serverless to Dedicated prior to early 2025, please see the following guide: [Convert a Serverless Instance to a Dedicated Cluster](https://www.mongodb.com/docs/atlas/tutorial/convert-serverless-to-dedicated/). **NOTE:** Manual migration can cause downtime and workload disruption.
```
Edit for conciseness.

---
page_title: "Migration Guide: Transition out of Serverless Instances and Shared-tier clusters"
---

# Migration Guide: Transition out of Serverless Instances and Shared-tier clusters

The goal of this guide is to help users transition from Serverless Instances and Shared-tier clusters (M2/M5) to Free, Flex or Dedicated Clusters. 

Starting in January 2025, all Shared-tier clusters (in both `mongodbatlas_cluster` and `mongodbatlas_advanced_cluster`) will automatically convert to Flex clusters. Similarly, in March 2025 all Serverless instances (`mongodb_serverless_instance`) will be converted into Free/Flex/Dedicated clusters, [depending on your existing configuration](https://www.mongodb.com/docs/atlas/flex-migration/).
Some of these conversions will result in configuration drift in Terraform. 

You can migrate from Serverless instances and Shared-tier clusters manually before March 2025. 

**--> NOTE:** We recommend waiting until March 2025 or later for Serverless instances and Shared-tier clusters to autoconvert.

> If a Serverless instance has $0 MRR, it automatically converts into a Free cluster. Else, if it does not fit the constraints of a Flex cluster, it will convert into a Dedicated cluster, resulting in downtime and workload disruption. Otherwise, it will convert to a Flex cluster.

Reviewer 3: Should we move this right after line 9?

Author: Moved

            region_name           = "US_EAST_1"
        }
        termination_protection_enabled = true
    }
    ```
2.  Run `terraform apply` to create the new resource.
3. Migrate data from your Serverless Instance to the Flex cluster using `mongodump` and `mongostore`.

    Please see the following guide on how to retrieve data from one cluster and store it in another cluster: [Convert a Serverless Instance to a Dedicated Cluster](https://www.mongodb.com/docs/atlas/tutorial/convert-serverless-to-dedicated/)

    Verify that your data is present within the Flex cluster before proceeding.
4. Delete the Serverless Instance by running a destroy command against it:

    `terraform destroy -target=mongodbatlas_serverless_instance.this`

 5. You may now safely remove the resource block for the Serverless Instance from your `.tf` file.

## From Serverless to Dedicated 

**Please note your Serverless instance will need to migrate to Decidated if it meets the following requirements:**
- \>= 5GB of data
- needs privatelink or continuous backup
- \> 500 ops/sec consistently.

You cannot migrate from Serverless to Dedicated using the Terraform provider.

> ### Pre-Autoconversion Migration Procedure

Reviewer 3: Post before pre

Author: In this case it's recommended to use the UI migration tool, that's why I didn't move the post before the pre. UI tool will cause downtime but not a need to change connection strings, which is better than the auto-conversion
